Python/chapter04/p01_route_between_nodes/miguelHx.py

```diff
@@ -0,0 +1,202 @@
+"""Python Version 3.9.2
+4.1 - Route Between Nodes:
+Given a directed graph, design an algorithm to
+find out whether there is a route between two nodes.
+"""
+import unittest
+
+from collections import deque
+from dataclasses import dataclass
+from typing import List, Deque
+
+
+@dataclass
+class Graph:
+    nodes: 'List[Node]'
+
+    def print_graph(self):
+        for node in self.nodes:
+            node.print_children()
+
+    def reset_visited(self):
+        for node in self.nodes:
+            node.visited = False
+
+
+@dataclass
+class Node:
+    id: int
+    children: 'List[Node]'
+    visited: bool = False
+
+    def add_child(self, *nodes: 'Node'):
+        for node in nodes:
+            self.children.append(node)
+
+    def print_children(self):
+        message = f'Adjacency list for node ({self.id}): '
+        for child in self.children:
+            message += '{}, '.format(child.id)
+        print(message)
+
+    def __str__(self):
+        return f'Node ({self.id}), visited: {self.visited}'
+
+
+def dfs_search(root: Node) -> List[int]:
+    """Simple DFS.
+    takes in a root, returns a list
+    of ids of the sequence of visited
+    nodes.
+
+    Args:
+        root (Node): starting node
+
+    Returns:
+        List[int]: list of node IDs (i.e. [0, 1, 3])
+    """
+    output = []
+    if root is None:
+        return
+    # print(f'Visiting node ({root.id})')
+    root.visited = True
+    # print(root.children)
+    output.append(root.id)
+    for node in root.children:
+        if not node.visited:
+            output.extend(dfs_search(node))
+    return output
+
+def bfs_search(root: Node) -> List[int]:
+    """Simple BFS.
+    takes in a root, returns a list
+    of ids of the sequence of visited
+    nodes.
+
+    Args:
+        root (Node): starting node
+
+    Returns:
+        List[int]: List[int]: list of node IDs (i.e. [0, 1, 4])
+    """
+    visited_nodes: List[Node] = []
+    visited_nodes.append(root)
```

Comment:
```suggestion
    visited_list: List[int] = [root.id]
    visited: Set[int] = set([root.id])
```

that way the `Node.visited` boolean is not needed anymore, and L99 can return `visited_list` as-is.

Reply:
265cf29b04bfc1e75ac795c3422ee831d68fe02c
Node.visited no longer needed